src/Data/Poly/Internal/Dense/Laurent.hs

```diff
@@ -0,0 +1,284 @@
+  forM_ [0 .. lenXs - 1] $ \k ->
+    MG.unsafeWrite zs (fromIntegral yp + k) (mul yc $ G.unsafeIndex xs k)
+  G.unsafeFreeze zs
+{-# INLINABLE scaleInternal #-}
+
+-- | Create an identity polynomial.
+pattern X :: (Eq a, Num a, G.Vector v a, Eq (v a)) => Laurent v a
+pattern X <- ((==) var -> True)
+  where X = var
+
+var :: forall a v. (Eq a, Num a, G.Vector v a, Eq (v a)) => Laurent v a
+var
+  | (1 :: a) == 0 = Laurent 0 (Poly G.empty)
+  | otherwise     = Laurent 0 (Poly $ G.fromList [0, 1])
+{-# INLINE var #-}
+
+-- | Evaluate at a given point.
+--
+-- >>> eval (X^-2 + 1 :: UPoly Rational) 3
+-- 10
+eval :: (Num a, G.Vector v a, G.Vector v (Int, a), G.Vector v Int, Fractional a) => Laurent v a -> a -> a
+eval = substitute (*) recip
+{-# INLINE eval #-}
+
+--eval' :: (Semiring a, G.Vector v a) => Laurent v a -> a -> a
+--eval' = substitute' times
+-- {-# INLINE eval' #-}
+
+data StrictPair a b = !a :*: !b
+
+infixr 1 :*:
+
+fst' :: StrictPair a b -> a
+fst' (a :*: _) = a
+
+-- | Substitute another polynomial instead of 'X'.
+--
+-- >>> subst (X^2 + 1 :: UPoly Int) (X + 1 :: UPoly Int)
+-- 1 * X^2 + 2 * X + 2
+subst :: (Eq a, Integral a, G.Vector v a, G.Vector v (Int, a), G.Vector v Int) => Laurent v a -> Laurent v a -> Laurent v a
+subst = substitute (scale 0) (error "")-- (1 `quotLaurent`)
+{-# INLINE subst #-}
+
+substitute :: (Num b, G.Vector v a, G.Vector v (Int, a), G.Vector v Int)
+  => (a -> b -> b)
+  -> (b -> b)
+  -> Laurent v a
+  -> b
+  -> b
+substitute f inv (Laurent n (Poly cs)) x
+  | n >= 0 = fst' $
+      G.foldl' (\(acc :*: xn) cn -> acc + f cn xn :*: x * xn) (0 :*: 1) cs
+  | otherwise = fst' $
+      G.foldl' (\(acc :*: xn) (i, cn) -> acc + f cn xn :*: pow x i)
+                (0 :*: pow x n)
+                (G.zip (G.fromList [(n+1)..(n + G.basicLength cs)]) cs)
+  where
+    pow b i
+      | i < 0 = inv b ^ negate i
+      | otherwise = b ^ i
+{-# INLINE substitute #-}
+
+substitute' :: (G.Vector v a, G.Vector v (Int, a), G.Vector v Int, S.Ring b) => (a -> b -> b) -> Laurent v a -> b -> b
+substitute' f (Laurent n (Poly cs)) x
+  | n >= 0 = fst' $
+      G.foldl' (\(acc :*: xn) cn -> acc `plus` f cn xn :*: x `times` xn) (zero :*: one) cs
+  | otherwise = fst' $
+      G.foldl' (\(acc :*: xn) (i, cn) -> acc `plus` f cn xn :*: pow x i)
+                (zero :*: pow x n)
+                (G.zip (G.fromList [(n+1)..(n + G.basicLength cs)]) cs)
+  where
+    pow b i
+      | i < 0 = S.negate b S.^ negate i
+      | otherwise = b S.^ i
+{-# INLINE substitute' #-}
+-- | Return a leading power and coefficient of a non-zero Laurent polynomial.
+--
+-- >>> leading ((2 * X^-1 + 1) * (2 * X^-2 - 1) :: UPoly Int)
+-- Just (0,-1)
+-- >>> leading (0 :: UPoly Int)
+-- Nothing
+leading :: G.Vector v a => Laurent v a -> Maybe (Int, a)
+leading (Laurent n (Poly v))
+  | G.null v  = Nothing
+  | otherwise = Just (n + G.length v - 1, G.last v)
+
+-- | Exponentiation of an identity Laurent polynomial to a negative integer.
```

Comment:
What is the justification behind this partial function? There is `monomial` already.

Reply:
It's a convenient function to define monomials with negative exponents

Reply:
Well, `monomial` already covers negative exponents as well.